src/presign/participant.rs

```diff
@@ -186,10 +228,92 @@ impl PresignParticipant {
             ));
         }
 
-        Ok(ret_messages)
+        let mut broadcast_messages = self.broadcast(
+            rng,
+            &MessageType::Presign(PresignMessageType::RoundOneBroadcast),
+            serialize!(&r1_public_broadcast)?,
+            message.id(),
+            "PresignR1Ciphertexts",
+        )?;
+        out_messages.append(&mut broadcast_messages);
+
+        // Additionally, handle any round 1 messages which may have been received too early
+        let retrieved_messages = self.fetch_messages(
+            MessageType::Presign(PresignMessageType::RoundOne),
+            message.id(),
+        )?;
+        let mut presign_record_option = None;
+        for msg in retrieved_messages.iter() {
+            let (pr_option, mut r2_msg) = self.handle_round_one_msg(rng, msg, main_storage)?;
+            out_messages.append(&mut r2_msg);
+            presign_record_option = pr_option;
+        }
+
+        Ok((presign_record_option, out_messages))
+    }
+
+    fn handle_round_one_broadcast_msg<R: RngCore + CryptoRng>(
+        &mut self,
+        rng: &mut R,
+        broadcast_message: &BroadcastOutput,
+        main_storage: &Storage,
+    ) -> Result<(Option<PresignRecord>, Vec<Message>)> {
+        if broadcast_message.tag != "PresignR1Ciphertexts" {
+            return bail!("Incorrect tag for Presign R1 Broadcast!");
```

Comment:
nit: You don't need `return` with `bail!`.

Reply:
Changing this causes a compiler error for some reason. 
`type annotations needed
cannot infer type of the type parameter `T` declared on the enum `Result``
in errors.rs:58

Reply:
Huh, weird.

Reply:
I can quickly look into this today. I agree with Marsella, this should work.